buffer/src/main/java/io/netty/buffer/AbstractReferenceCountedByteBuf.java

```diff
@@ -113,18 +130,38 @@ public boolean release(int decrement) {
     }
 
     private boolean release0(int decrement) {
-        int oldRef = refCntUpdater.getAndAdd(this, -decrement);
-        if (oldRef == decrement) {
-            deallocate();
-            return true;
-        }
-        if (oldRef < decrement || oldRef - decrement > oldRef) {
-            // Ensure we don't over-release, and avoid underflow.
-            refCntUpdater.getAndAdd(this, decrement);
-            throw new IllegalReferenceCountException(oldRef, -decrement);
+        int rawCnt = nonVolatileRawCnt();
+        for (boolean firstTry = true;; firstTry = false) {
+            if ((rawCnt & 1) != 0) {
+                throw new IllegalReferenceCountException(0, -decrement);
+            }
+            int realCnt = rawCnt >>> 1;
+            if (decrement == realCnt) {
+                // most likely case
+                if (refCntUpdater.compareAndSet(this, rawCnt, 1)) { // any odd number
```

Comment:
optional: this could just be getAndAdd(1) right?

Reply:
@carl-mastrangelo I think the compare is needed here since the count may have increased in which case we don't want the release to succeed (as soon as the ref changes to an odd number all other releases/retains will fail)